openedx/core/djangoapps/content_libraries/models.py

```diff
@@ -127,3 +161,82 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
     def __str__(self):
         who = self.user.username if self.user else self.group.name
         return f"ContentLibraryPermission ({self.access_level} for {who})"
+
+
+class LtiProfileManager(models.Manager):
+    """Custom manager of LtiProfile mode."""
+
+    def get_from_claims(self, *, iss, sub):
+        """Get the LtiProfile from a ISS and SUB claims."""
+        return self.get(platform_id=iss, subject_id=sub)
+
+    def get_or_create_from_claims(self, *, iss, sub):
+        try:
+            return self.get_from_claims(iss=iss, sub=sub)
+        except self.model.DoesNotExist:
+            # User will be created on ``save()``.
+            return self.create(platform_id=iss, subject_id=sub)
+
+
+class LtiProfile(models.Model):
+    """
+    Content Librarie LTI's profile for Open edX users.
+
+    Unless Anonymous, this should be a unique representation of the LTI subject
+    (as per the client token ``sub`` identify claim) that initiated an LTI
+    launch through Content Libraries.
+    """
+
+    objects = LtiProfileManager()
+
+    user = models.OneToOneField(
+        get_user_model(),
+        null=True,
+        on_delete=models.CASCADE,
+        related_name='contentlibraries_ltiprofile',
+        verbose_name=_('open edx user'),
+    )
+
+    platform_id = models.CharField(
+        max_length=255,
+        verbose_name=_('platform identifier'),
+        help_text=_("The LTI platform identifier to which this profile belongs "
+                    "to.")
+    )
+
+    subject_id = models.CharField(
+        max_length=255,
+        verbose_name=_('subject identifier'),
+        help_text=_('Identifies the entity that initiated the deep linking '
+                    'request, commonly a user.  If set to ``None`` the profile '
+                    'belongs to the Anonymous entity.')
+    )
+
+    created_at = models.DateTimeField(
+        auto_now_add=True
+    )
+
+    class Meta:
+        unique_together = ['platform_id', 'subject_id']
+
+    @property
+    def subject_url(self):
+        """An local URL that is known to uniquely identify this profile."""
+        return f'{self.platform_id}#subject={self.subject_id}'
+
+    def save(self, **kwds):
+        """Get or create an edx user on save."""
+        if not self.user:
+            uid = uuid.uuid5(uuid.NAMESPACE_URL, self.subject_url)
+            username = f'urn:openedx:content_libraries:username:{uid}'
+            email = f'{uid}@{ContentLibrariesConfig.name}'
```

Comment:
Are you talking about the username scheme or creating users automatically on `save()`? Or both?

Regarding the username, [one alternative approach is creating randomly generated usernames][1]. Both would work seamlessly, but I believe it is more maintainable to have them namespaced with some scheme, hence the patch as is. The primary benefit is enforcement of LTI 1.3's premise that ISS/SUB claims are unique per LTI integration at the authentication level. It means the username inside Open EdX instances will have a 1:1 mapping with the LTI platform for a given deployment ID.

Drawback is we don't have additional logic to block the creation of users prefixed with `urn: openedx` (or whatever namespace we want to incorporate as reserved). Or protection against the very unlikely event of some instance somewhere already using this prefix.

[1]: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/lti_provider/users.py#L46

Reply:
Not sure if this is covered/flagged already but we limit PII sharing if a certain LTI waffle flag isn't set or enabled per course currently. I don't know how this would / could extend to this broader blockstore configuration concept, but ideally, we by default aren't using learner PII for any automated IDs unless that sharing is explicitly set/enabled?

Reply:
@marcotuts:

> we by default aren't using learner PII for any automated IDs unless that sharing is explicitly set/enabled?

I don't think that restriction applies here. The LTI launches performed through Content Libraries are detached from any current user in the platform. Except for the `subject_url` I think all the other attributes here are dynamically created. `subject_url` is supposed to be opaque, and even so, we "hash it" under an UUIDv5.

Do you agree? If not, any suggestion on  what would be the next steps here?

Reply:
> The LTI launches performed through Content Libraries are detached from any current user in the platform.

What if the user is already logged in to Open edX, and then does an LTI launch in an iframe - will it use their existing account? From the code it sounds like yes ("Unless anonymous..."), and I guess that's what I would expect but I'm a little unclear on this point.

> We'll probably need to check with edX that this specific way of auto-creating users works for them. No action needed now, though - just a note for the future.

Yes, we'll need to figure that out now. The concern that I've seen expressed from devops folks in the past is about the users table becoming filled with tens or hundreds thousands of user accounts from individual LTI sessions. Or to be specific, in the past the concern was about different implementations of "anonymous access" to courses generating tons of user account detritus I think, but it's very similar in practice.

That's why with the blockstore runtime, I implemented support for "anonymous users" that doesn't create `User` objects in the database at all, which [was an explicit design goal](https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/xblock#high-level-design-goals): 

> To provide first-class support for anonymous usage of XBlocks (use by unregistered users): If the user is unregistered, field data will be saved into the django cache rather than the database.

Question 1: is it possible to use the new runtime's existing approach to "anonymous users" with LTI, or is it incompatible?
Question 2: how does this LTI tool implementation compare with the platform's existing (course/modulestore) LTI tool provider implementation in terms of creating users?
Question 3: If we need to create "real" user entries for LTI users, as done in this PR, can we create a cleanup task that will periodically delete them? Or would that erase important data?

See 

https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/utils.py#L143-L152

and

https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/runtime/ephemeral_field_data.py#L1-L17

Reply:
> > The LTI launches performed through Content Libraries are detached from any
> > current user in the platform.
> 
> What if the user is already logged in to Open edX, and then does an LTI launch
> in an iframe - will it use their existing account? From the code it sounds
> like yes ("Unless anonymous..."), and I guess that's what I would expect but
> I'm a little unclear on this point.

As a matter of fact, no. It won't.  First, because I think the Platform's
cookies are not cross-site (I think they have `SameSite=Lax`, but I would have
to double-check), but more importantly, LTI launches ignore whatever user in the
current session and perform the whole auth/login cycle based on the claims
passed from the platform, in particular, the [user identify claim
`sub`](http://www.imsglobal.org/spec/lti/v1p3/#user-identity-claims).

> > We'll probably need to check with edX that this specific way of
> > auto-creating users works for them. No action needed now, though - just a
> > note for the future.
> 
> Yes, we'll need to figure that out now. The concern that I've seen expressed
> from devops folks in the past is about the users table becoming filled with
> tens or hundreds thousands of user accounts from individual LTI sessions. Or
> to be specific, in the past the concern was about different implementations of
> "anonymous access" to courses generating tons of user account detritus I
> think, but it's very similar in practice.

That is a valid concern.  I pretty much followed the current LTI 1.1
implementation footsteps, and created users attached to custom LTI profiles.
This turned out to be comfortable with the current grading signals that
identify responses per `(user_id, usage_id)`.  By creating one user per LTI
launch, [I was able to pull the proper claims to reply to the LTI Platform to
sync grades](https://github.com/open-craft/edx-platform/blob/35c24c884922f6899955b9b79b3c3983b19e1800/openedx/core/djangoapps/content_libraries/signal_handlers.py#L34).

> That's why with the blockstore runtime, I implemented support for "anonymous
> users" that doesn't create `User` objects in the database at all, which [was
> an explicit design
> goal](https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/xblock#high-level-design-goals):
> 
> > To provide first-class support for anonymous usage of XBlocks (use by
> > unregistered users): If the user is unregistered, field data will be saved
> > into the django cache rather than the database.

Oh, that's interesting.  With this approach, how the grading signal handlers
would be called?

> Question 1: is it possible to use the new runtime's existing approach to
> "anonymous users" with LTI, or is it incompatible?

Not sure. If the cached user "entity" is dropped, another launch on the
same `SUB` and `ISS` will re-populate the relationships between user id and the
graded resource? For me to continue to handle grade signals properly, that
would be necessary.  I would need to look at the code and perform
some tests here.

> Question 2: how does this LTI tool implementation compare with the platform's
> existing (course/modulestore) LTI tool provider implementation in terms of
> creating users?

It's pretty much the same approach.  We [clarified that in the
ADR](https://github.com/edx/edx-platform/pull/27089/files).

> Question 3: If we need to create "real" user entries for LTI users, as done in
> this PR, can we create a cleanup task that will periodically delete them? Or
> would that erase important data?

From this patch perspective, no.  I am not sure if the grading subsystem will
have any side-effect if we delete the user associate with some of the grades.

Reply:
> > support for "anonymous users" that doesn't create User objects in the database at all, which was an explicit design goal
>
> Oh, that's interesting. With this approach, how the grading signal handlers
would be called?

Currently, [they are not](https://github.com/edx/edx-platform/blob/91d5b668a895c89d695227b48afe2c2552c284e2/openedx/core/djangoapps/xblock/runtime/runtime.py#L167-L168) (if the user is anonymous), for that exact reason - I didn't want to pass a fake user object into the existing signal handlers as that could break their API. So perhaps that answers my question: no the ephemeral users (not saved into MySQL) probably won't work for LTI if the implementation is based on the existing Django signal handlers. 

That said, one could very easily modify [this code](https://github.com/edx/edx-platform/blob/91d5b668a895c89d695227b48afe2c2552c284e2/openedx/core/djangoapps/xblock/runtime/runtime.py#L167-L168) to emit a second signal like `ANONYMOUS_USER_SCORE_PUBLISHED` and `ANONYMOUS_PROBLEM_WEIGHTED_SCORE_CHANGED` which the LTI code could listen to, associate with a user, and then send back to the LTI client. If LTI users and grades don't need to persist in the platform for longer than a few hours, I think this would be a great approach, and avoid polluting the `User` table with temporary users. But if you need to be able to retrieve the same LTI user/profile days later, then the approach you have already taken is better/necessary.

> > Question 3: If we need to create "real" user entries for LTI users, as done in
this PR, can we create a cleanup task that will periodically delete them? Or
would that erase important data?
>
> From this patch perspective, no. I am not sure if the grading subsystem will
have any side-effect if we delete the user associate with some of the grades.

I meant as a future addition, not in this PR. And I was assuming it would delete the users and grades, i.e. do a cascading user delete.

For now, can you please just make sure there is some easy way to select all LTI 1.3 users in the database, so that doing a mass delete of them (or a mass delete of "all LTI users older than 1 month", etc.) is at least theoretically possible? It seems what you have here will already permit that, I just want to be explicit.